core/src/main/java/org/infinispan/util/ByteString.java

```diff
@@ -0,0 +1,90 @@
+package org.infinispan.util;
+
+import org.infinispan.commons.marshall.AdvancedExternalizer;
+import org.infinispan.marshall.core.Ids;
+
+import java.io.IOException;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.nio.charset.Charset;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Set;
+
+/**
+ * @author Tristan Tarrant
+ * @since 9.0
+ */
+public class ByteString {
+   public static final Charset CHARSET = Charset.forName("UTF-8");
+   private static final ByteString EMPTY = new ByteString(new byte[0]);
+   private final byte[] b;
+   private final transient int hash;
+
+   ByteString(byte[] b) {
+      this.b = b;
+      this.hash = Arrays.hashCode(b);
+   }
+
+   public static ByteString fromString(String s) {
+      if (s.length() == 0)
+         return EMPTY;
+      else
+         return new ByteString(s.getBytes(CHARSET));
+   }
+
+   public static ByteString emptyString() {
+      return EMPTY;
+   }
+
+   @Override
+   public int hashCode() {
+      return hash;
+   }
+
+   @Override
+   public boolean equals(Object o) {
+      if (this == o) return true;
+      if (o == null || getClass() != o.getClass()) return false;
+      ByteString that = (ByteString) o;
+      return Arrays.equals(b, that.b);
+   }
+
+   @Override
+   public String toString() {
+      return new String(b, CHARSET);
+   }
+
+   public static class Externalizer implements AdvancedExternalizer<ByteString> {
+
+      @Override
+      public Set<Class<? extends ByteString>> getTypeClasses() {
+         return Collections.singleton(ByteString.class);
+      }
+
+      @Override
+      public Integer getId() {
+         return Ids.ASCII_STRING;
+      }
+
+      @Override
+      public void writeObject(ObjectOutput output, ByteString object) throws IOException {
+         output.writeByte(object.b.length);
+         if (object.b.length > 0) {
+            output.write(object.b);
+         }
+      }
+
+      @Override
+      public ByteString readObject(ObjectInput input) throws IOException, ClassNotFoundException {
+         short l = input.readByte();
```

Comment:
Could we use `len` here or something more descriptive? `l` looks like '1' and it doesn't look good in `new byte[1]` :)